ddtrace/tracer/option.go

```diff
@@ -138,6 +138,24 @@ type config struct {
 
 	// enabled reports whether tracing is enabled.
 	enabled bool
+
+	// postProcessor holds the function used to process finished spans of a trace.
+	// It reports whether the trace should be dropped.
+	postProcessor func([]ddtrace.ReadWriteSpan) bool
+}
+
+// WithPostProcessor enables processing finished spans of a trace by f. Spans
+// can be read from and written to in the processor by using the methods specified by
+// the ReadWriteSpan interface. f reports whether the trace should be dropped (dropped
+// traces and spans will not be accounted for in APM stats).
+//
+// Please note that modifying any of the ReadWriteSpans outside the scope of the processor
+// function can and will likely cause a panic, and that using the processor will incur additional
+// overhead (which is directly linked to the performance of the code in f).
+func WithPostProcessor(f func([]ddtrace.ReadWriteSpan) bool) StartOption {
```

Comment:
I was thinking about this `bool` return type. If I understand the code correctly, returning false guarantees that the trace will be dropped. However, returning true doesn't guarantee that the trace will be kept, as it could be sampled later. Is that correct?

This also isn't descriptive on it's own. Meaning that `WithPostProcessor` returning true doesn't mean anything to me until I read the docs. Ideally the API can speak for itself as much as possible.

Do we need a way of allowing users to 1. Force Drop 2. Force Keep 3. Unaffected by post processor ?

I wonder if it might be better to return an enum of something like this:
```
type SamplingDecision int64
const (
	Unaffected SamplingDecision = iota
	ForceDrop
	ForceKeep
)
```

(Sorry for not considering this when you were designing it, but better to discuss this now rather than after we've already merged this)

Reply:
> If I understand the code correctly, returning false guarantees that the trace will be dropped. However, returning true doesn't guarantee that the trace will be kept, as it could be sampled later. Is that correct?

Yes, this is correct. The trace could still be sampled out in the trace-agent after this.

`1. Force Drop`  and `3. Unaffected by post processor` is currently what is allowed by the processor. 

For  allowing `2. Force Keep`, if we do support this in the processor, I believe the cleanest way would be to integrate this with how setting sampling priority currently works (eg. setting `ext.ManualKeep`, method `setSamplingPriorityLocked`...etc). 

If we allow the user to modify the sampling priority in the processor using the current methods, there are some implications:
- Setting `ext.ManualKeep` propagates this decision to downstream services. It would not be possible to make this work when modifying sampling priority in the processor.
- Doing so would required _some_ adaptations to the sampling methods, because setting sampling priority of a span is currently assumed to be done on spans that are not finished, and the sampling priority of a trace get's locked (not actual mutex, just a field) once this is set. 
The adaptations would not change how sampling works without a processor, but the code will need to be updated in a way that also fits in with setting sampling priority in the processor, which will increase the complexity of it.
- As we are adding new samplers, it may be harder to maintain this if we integrate it with the processor.

**TLDR:** It is possible to allow users to modify the sampling priority in the processor, but there is a trade-off between the risk increasing the complexity of the code/ making it harder to maintain vs not allowing changing sampling priority (technically, customer's will still get half of the equation, as they can force drop but not force keep).

Does this make sense ? What are your thoughts on this ?

cc @gbbr if you have thoughts on this as we had discussed this on slack.

Reply:
I'd say to not complicate things too much, there are already multiple ways to express decisions in the code, sampling priorities, tags, etc. I'd leave as it is because this is what was requested by users and reflects the behaviour of other tracers too.

This was just a quick answer in order to unblock but if you need me to go into more detail please let me know and I'm happy to provide.

Reply:
Thanks for thinking about this and writing that out. Can you link to where other language tracers return a bool in this way?
For example, it looks like Java lets you modify the spans, but I don't see them letting you drop the span at the same time (https://javadoc.io/static/com.datadoghq/dd-trace-api/0.104.0/datadog/trace/api/interceptor/TraceInterceptor.html)

Reply:
I always thought it was quite common for a filter function to return a boolean representing whether that element stays or goes...

I'll leave it up to you, but I'd advise that if you do end up using a type, to try and avoid the word "sampling" because sampling is a very complex topic with many mechanisms implemented (local rate sampler, priority sampler, rule sampler, single span sampler, etc) and if we add one more it will end up confusing the user.

Reply:
Do you have examples (especially of Go packages) that return a bool from a filter function? Not because I doubt you, but because it would be helpful to see those examples as a reference (e.g. how do they document it, how do they name it, do they use a named type).
You also mentioned that other tracers do this too, but I couldn't find examples. Do you have links we can reference for the same reason?

Reply:
One other way to make this clearer if we keep the bool return would be to name the return value, e.g.
```
func WithPostProcessor(f func([]ddtrace.ReadWriteSpan) (dropTrace bool)) StartOption {
```
That being said, this does not solve the _shoehorn ourselves into this kind of return value that can't change or be flexible in the future_  issue mentioned by Katie above.

The idea of using an enum also sounds good to me, and could look something like:
```
// Need to find a better type name.
type ProcessorDecision int64
const (
	Unaffected ProcessorDecision = iota
	ForceDrop
)
```

Name `ProcessorDecision` isn't very explicit, but I can't think of a good name for this at the moment. Will keep thinking about this and find a better name.

Let me know what you think ?

Reply:
@katiehockman some things that come to mind are:

* [ForEach functions](https://cs.github.com/?q=foreach%20language%3Ago) which quite commonly report whether to continue or not. 
* The commonly returned `ok` boolean argument returned by many functions which reports whether something is good to continue with or not.
* If I search all Go code on Github for the word `filter` I find [plenty of examples](https://cs.github.com/?scopeName=All+repos&scope=&q=filter+language%3Ago) using it like that.
* Also, I'm probably biased by all my past Javascript writing, where all [filters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) work like that.

I don't have any processor examples from other tracers, it was just an assumption I made based on what I (maybe mis-) understood from other discussions.

As you have said in the past yourself, I'd be wary of adding more API into the mix. I'm thinking of situations where people are trying to get an overview of the tracer API. An additional type increases confusion and complexity and its only reason is there to server a post-processor which is likely to not be used very widely because of performance implications. Is it worth it?

If you feel strongly about this, do it, but to me it doesn't make sense to have a return type when the decision is strictly binary (= boolean).

Reply:
As far as bool versus named type, I don't feel all that strongly. If we're confident that this API won't need to change in the future (i.e. "drop" and "unaffected" are the only two that we will care about), then naming the variable in the definition may be sufficient. Then it can stay a bool.

> If I search all Go code on Github for the word filter I find [plenty of examples](https://cs.github.com/?scopeName=All+repos&scope=&q=filter+language%3Ago) using it like that.

Right, but I usually see a bool used something like this in that case:

```
func Filter(s []interface{}, f func(interface{}) bool) []interface{}
```

This isn't really a filter though, is it? You can't filter out individual spans in the trace using this API. It's either keep the whole trace, or drop the whole trace. Or am I mistaken there?
Additionally, returning `true` in our API means the whole trace should be dropped, but that's the opposite meaning to a filter. https://pkg.go.dev/go/ast#FilterDecl for example, returns `true` if any items *remain*. We return `true` if *nothing* should remain.

I think the part I've been hung up on is that this API is designed to do two things:
1. Apply changes to individual spans in a trace.
2. Drop the entire trace based on properties of its spans.

The name "`WithPostProcessor`" only covers (1). It's not clear from the name that it can do (2).
I'm really struggling to find a name that covers both things though. Maybe the named variable is the clearest option at this point.

Reply:
> The name "WithPostProcessor" only covers (1). It's not clear from the name that it can do (2).
I'm really struggling to find a name that covers both things though. Maybe the named variable is the clearest option at this point.

The named variable + the documentation should help. Sometimes it's not easy to find a name which clearly represents everything a piece of API does. You'd expect users however to read the documentation for API that they are using 😄

Reply:
Agreed that the named variable is the clearest option at this point. 

If ever we need to change the API to allow users to force keep, we will be able to do so by letting the user modify the sampling priority in the processor (e.g. setting `ext.ManualKeep`).

If we go with the named variable: Currently, the processor function returns `true` if the trace is unaffected and `false` if the trace should be dropped. Because of this, naming the variable is tough. We can't call it `shouldKeep` because there is no guarantee it will be kept in the agent, and naming it `shouldNotDrop` is also strange. 

I suggest inverting the return statement to return true if the trace should be dropped, and false if it shouldn't. If we do this, the variable would be named `shouldDrop`. In my opinion, this would be the clearest:
```
func WithPostProcessor(f func([]ddtrace.ReadWriteSpan) (shouldDrop bool)) StartOption {
```

Thoughts ?

Reply:
My instinct in this case would be that the default value of a bool should mean that nothing changes as a result of this code. So that would mean that `false` would mean that the trace sampling is unaffected. `true` would indicate that some action should be taken, and thus the trace should be dropped. So yes I agree.

I'd go with:
`func WithPostProcessor(f func([]ddtrace.ReadWriteSpan) (dropTrace bool)) StartOption {`

Reply:
The variable is now named `dropTrace` 🙂. Thanks for all the input here! Marking this conversation as resolved.